File: exam_LRX/py/work8.py
# 8.给出图像增强处理中的空间域增强的加法、减法和乘法的实例
import cv2
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

# 设置中文字体
rcParams['font.family'] = 'SimHei'  # 设置为黑体
rcParams['axes.unicode_minus'] = False  # 解决负号显示问题


def enhance_image_addition(image, constant):
    """加法增强：将常数加到每个像素值"""
    enhanced_image = cv2.add(image, constant)
    return enhanced_image


def enhance_image_subtraction(image, constant):
    """减法增强：从每个像素值中减去常数"""
    enhanced_image = cv2.subtract(image, constant)
    return enhanced_image


def enhance_image_multiplication(image, constant):
    """乘法增强：将每个像素值与常数相乘"""
    enhanced_image = cv2.multiply(image, constant)
    return enhanced_image


def display_images_in_one_window(original_image, enhanced_addition, enhanced_subtraction, enhanced_multiplication):
    """在同一个窗口中显示原图和所有增强图像"""
    plt.figure(figsize=(12, 8))

    # 原图
    plt.subplot(2, 2, 1)
    plt.imshow(original_image, cmap='gray')
    plt.title('原始图像')
    plt.axis('off')

    # 加法增强
    plt.subplot(2, 2, 2)
    plt.imshow(enhanced_addition, cmap='gray')
    plt.title('加法增强')
    plt.axis('off')

    # 减法增强
    plt.subplot(2, 2, 3)
    plt.imshow(enhanced_subtraction, cmap='gray')
    plt.title('减法增强')
    plt.axis('off')

    # 乘法增强
    plt.subplot(2, 2, 4)
    plt.imshow(enhanced_multiplication, cmap='gray')
    plt.title('乘法增强')
    plt.axis('off')

    plt.tight_layout()
    plt.show()


def main():
    # 读取图像
    image = cv2.imread('../image/image9.png', cv2.IMREAD_GRAYSCALE)

    if image is None:
        print("图像加载失败，请确保文件路径正确。")
        return

    # 选择常数值
    constant = 50

    # 加法增强
    enhanced_addition = enhance_image_addition(image, constant)

    # 减法增强
    enhanced_subtraction = enhance_image_subtraction(image, constant)

    # 乘法增强
    constant = 1.5  # 修改常数值以增强乘法效果
    enhanced_multiplication = enhance_image_multiplication(image, constant)

    # 在同一个窗口内显示原图和所有增强后的图像
    display_images_in_one_window(image, enhanced_addition, enhanced_subtraction, enhanced_multiplication)


if __name__ == "__main__":
    main()

File: exam_LRX/cpp/work8.cpp
//
// Created by liryi on 24-11-22.
// 8.给出图像增强处理中的空间域增强的加法、减法和乘法的实例
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

// 加法增强
Mat enhanceImageAddition(const Mat& image, int constant) {
    Mat enhancedImage;
    add(image, Scalar(constant), enhancedImage);
    return enhancedImage;
}

// 减法增强
Mat enhanceImageSubtraction(const Mat& image, int constant) {
    Mat enhancedImage;
    subtract(image, Scalar(constant), enhancedImage);
    return enhancedImage;
}

// 乘法增强
Mat enhanceImageMultiplication(const Mat& image, double constant) {
    Mat enhancedImage;
    multiply(image, Scalar(constant), enhancedImage);
    return enhancedImage;
}

// 在一个窗口中显示多张图片
void displayImagesInOneWindow(const Mat& originalImage, const Mat& enhancedAddition,
                              const Mat& enhancedSubtraction, const Mat& enhancedMultiplication) {
    // 调整大小和布局
    int width = originalImage.cols;
    int height = originalImage.rows;
    Mat display = Mat::zeros(height * 2, width * 2, originalImage.type());

    // 放置原始图像
    originalImage.copyTo(display(Rect(0, 0, width, height)));
    putText(display, "Original image", Point(10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置加法增强
    enhancedAddition.copyTo(display(Rect(width, 0, width, height)));
    putText(display, "Addition enhancement", Point(width + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置减法增强
    enhancedSubtraction.copyTo(display(Rect(0, height, width, height)));
    putText(display, "Subtractive enhancement", Point(10, height + 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置乘法增强
    enhancedMultiplication.copyTo(display(Rect(width, height, width, height)));
    putText(display, "Multiplication enhancement", Point(width + 10, height + 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 显示结果
    namedWindow("Enhancement effect comparison", WINDOW_NORMAL);
    imshow("Enhancement effect comparison", display);
    waitKey(0);
}

int main() {
    // 读取图像
    string imagePath = "E:\\DL_class\\exam_LRX\\cpp\\images\\image9.png"; // 修改为你的图像路径
    Mat image = imread(imagePath, IMREAD_GRAYSCALE);

    if (image.empty()) {
        cout << "imager error" << endl;
        return -1;
    }

    // 常数值
    int constantAddSub = 50;
    double constantMul = 1.5;

    // 加法增强
    Mat enhancedAddition = enhanceImageAddition(image, constantAddSub);

    // 减法增强
    Mat enhancedSubtraction = enhanceImageSubtraction(image, constantAddSub);

    // 乘法增强
    Mat enhancedMultiplication = enhanceImageMultiplication(image, constantMul);

    // 显示所有结果
    displayImagesInOneWindow(image, enhancedAddition, enhancedSubtraction, enhancedMultiplication);

    return 0;
}

